File: LeetCode/LeetCode_916_Word_Subsets.py
from typing import List 
def wordSubsets(words1: List[str], words2: List[str]) -> List[str]:
        word1hash = {}
        for word in words1 : 
            word1hash[word] = {}
            for letter in word :
                word1hash[word].setdefault(letter,0)
                word1hash[word][letter] += 1
        
        words2hash = {}
        for word in words2 : 
            words2hash[word] = {}
            for letter in word : 
                words2hash[word].setdefault(letter,0)
                words2hash[word][letter] += 1
        
        result = []
        for word , hash1 in word1hash.items() :
            
            isUniversal = True
            for compareword, hash2 in words2hash.items() :
                # print(word,word1hash[word],"\n",compareword,hash2, sep=" ")
                
                for letter, freq in hash2.items() :
                    if letter not in hash1.keys() : 
                        isUniversal = False 
                        break
                    # print(hash1[letter], freq)
                    if hash1[letter] < freq :
                        isUniversal = False
                        break 
                if not isUniversal : break
            if isUniversal :
                result.append(word)


        return result


words1 = input().split()
words2 = input().split()

result = wordSubsets(words1, words2)
print(*result)

File: LeetCode/LeetCode_916_Word_Subsets.cpp
#include<bits/stdc++.h>
using namespace std ;

vector<string> wordSubsets(vector<string>& words1, vector<string>& words2) {
    vector<string> result ;
    map<string, vector<int>> hash1 , hash2; 


    for(string word : words1) {
        hash1[word] = vector<int> (26,0) ;
        for (char letter : word) {
            hash1[word]['a'-letter] += 1;
        }
    }

    for(string word : words2) {
        hash2[word] = vector<int> (26,0) ;
        for (char letter : word) {
            hash2[word]['a'-letter] += 1;
        }
    }

    for(string word : words1) {
        bool isUniversal = true ;
        for (auto & pr : hash2) {
            string word = pr.first ;
            vector<int> hash = pr.second ;

            for (int i = 0; i < 26; i++)
            {
                if (hash[i] < hash1[word][i]) {
                    isUniversal = false ;
                    break;
                }
            }
            
            if(isUniversal) {
                result.push_back(word);
            }

        }
    }

    return result ;
}
int main() {
    int words1num, words2num ;
    cin >> words1num >> words2num ;

    vector<string> words1, words2 ;
    for (int i = 0; i < words1num; i++)
    {
        string word ;
        cin >> word ;
        // cout << word ;
        words1.push_back(word);
    }

    for (int i = 0; i < words2num; i++)
    {
        string word ;
        cin >> word ;
        words2.push_back(word);
    }
    
    vector<string> result = wordSubsets(words1, words2) ;
    for(string word : words2) {
        cout << word << " " ;
    }
    
    return 0 ;
}